### PR TITLE
V2 first run

### PR DESCRIPTION
**Pull request by Skrilltrax**
## :loudspeaker: Type of change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring

## :scroll: Description
Use livedata for app state and add first run flow

## :bulb: Motivation and Context
v2

## :green_heart: How did you test it?
App builds and works as expected.

## :pencil: Checklist

- [x] I ran `./gradlew spotlessApply` before submitting the PR
- [x] I reviewed submitted code

### CODE REVIEW

File: app/src/main/kotlin/com/passwordstore/android/ui/activity/SplashActivity.kt [hunk 1]
```diff
@@ -4,18 +4,35 @@
...
 
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
+        prefs = PreferenceManager.getDefaultSharedPreferences(this)
         binding = ActivitySplashBinding.inflate(layoutInflater)
         EdgeToEdge.setUpRoot(binding.root as ViewGroup)
         setContentView(binding.root)
+        if (prefs.getBoolean(IS_FIRST_RUN, true)) {
+            startFirstRunFlow()
+        } else {
```
Comment (msfjarvis) [inline app/src/main/kotlin/com/passwordstore/android/ui/activity/SplashActivity.kt]: unneeded

File: app/src/main/kotlin/com/passwordstore/android/ui/activity/BaseActivity.kt [hunk 2]
```diff
@@ -12,13 +12,15 @@ import com.passwordstore.android.utils.Authenticator
 
 abstract class BaseActivity : AppCompatActivity() {
 
+    private val passwordStoreApplication by lazy { application as PasswordStoreApplication }
```
Comment (msfjarvis) [inline app/src/main/kotlin/com/passwordstore/android/ui/activity/BaseActivity.kt]: This variable name is far too long. Abbreviate it to just `app` or `application` depending on your thoughts about name shadowing.

File: app/src/main/kotlin/com/passwordstore/android/ui/activity/FirstRunActivity.kt [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+package com.passwordstore.android.ui.activity
+
+import android.os.Bundle
+import android.view.ViewGroup
+import com.passwordstore.android.databinding.ActivityFirstRunBinding
+import com.passwordstore.android.ui.EdgeToEdge
+
+class FirstRunActivity : BaseActivity() {
+
+    private lateinit var binding: ActivityFirstRunBinding
```
Comment (msfjarvis) [inline app/src/main/kotlin/com/passwordstore/android/ui/activity/FirstRunActivity.kt]: Don't make this a class variable when we're not using it outside `onCreate`

File: app/src/main/kotlin/com/passwordstore/android/ui/activity/SplashActivity.kt [hunk 4]
```diff
@@ -4,18 +4,35 @@
...
+import com.passwordstore.android.utils.Constants.IS_FIRST_RUN
 
 class SplashActivity : BaseActivity() {
+
     private lateinit var binding: ActivitySplashBinding
+    private lateinit var prefs: SharedPreferences
 
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
+        prefs = PreferenceManager.getDefaultSharedPreferences(this)
```
Comment (msfjarvis) [inline app/src/main/kotlin/com/passwordstore/android/ui/activity/SplashActivity.kt]: Inject this through DI. I kinda wanna explore Dagger now that I migrated Viscerion (https://github.com/msfjarvis/viscerion/pull/214) and have a better understanding of everything.

File: app/src/main/kotlin/com/passwordstore/android/utils/Constants.kt [hunk 5]
```diff
@@ -0,0 +1,9 @@
+/*
+ * Copyright © 2014-2019 The Android Password Store Authors. All Rights Reserved.
+ * SPDX-License-Identifier: GPL-3.0-only
+ */
+package com.passwordstore.android.utils
+
+object Constants {
+    const val IS_FIRST_RUN = "is_first_run"
```
Comment (msfjarvis) [inline app/src/main/kotlin/com/passwordstore/android/utils/Constants.kt]: This is a preference key, right? Rename the class to match that, something like `PrefKeys` or `PreferenceKeys` alongwith a comment at the top of the class that makes it absolutely clear for any future contributors.

File: app/src/main/kotlin/com/passwordstore/android/ui/activity/SplashActivity.kt [hunk 6]
```diff
@@ -4,18 +4,35 @@
...
+        if (prefs.getBoolean(IS_FIRST_RUN, true)) {
+            startFirstRunFlow()
+        } else {
+        }
+    }
+
+    private fun startFirstRunFlow() {
+        startActivity(Intent(this, FirstRunActivity::class.java))
+        overridePendingTransition(android.R.anim.fade_in, android.R.anim.fade_out)
+        finish()
```
Comment (msfjarvis) [inline app/src/main/kotlin/com/passwordstore/android/ui/activity/SplashActivity.kt]: Don't you wanna change `IS_FIRST_RUN` to false somewhere here?

File: app/src/main/res/layout/activity_first_run.xml [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+        android:text="@string/app_name"
+        android:textColor="@color/color_control_normal"
+        android:textStyle="bold"
+        android:layout_centerHorizontal="true"
+        app:layout_constraintTop_toBottomOf="@+id/app_icon"
+        app:layout_constraintStart_toStartOf="parent"
+        app:layout_constraintEnd_toEndOf="parent" />
+
+    <com.google.android.material.button.MaterialButton
+        style="@style/Widget.AppCompat.Button.Colored"
```
Comment (msfjarvis) [inline app/src/main/res/layout/activity_first_run.xml]: Any reason to prefer `AppCompat` over `MaterialComponents`?

File: app/src/main/res/values/styles.xml [hunk 8]
```diff
@@ -19,8 +19,6 @@
         <item name="android:textColor">@color/primary_text_color</item>
         <item name="android:statusBarColor">@color/primary_dark_color</item>
         <item name="android:navigationBarColor">@android:color/transparent</item>
-        <item name="android:windowLightStatusBar" tools:targetApi="m">@bool/light_status_bar</item>
```
Comment (msfjarvis) [inline app/src/main/res/values/styles.xml]: `#ytho`

File: app/src/main/kotlin/com/passwordstore/android/ui/activity/SplashActivity.kt (hunk 1 above)
Comment (Skrilltrax) [reply app/src/main/kotlin/com/passwordstore/android/ui/activity/SplashActivity.kt]: Which part ?

File: app/src/main/kotlin/com/passwordstore/android/ui/activity/SplashActivity.kt (hunk 4 above)
Comment (Skrilltrax) [reply app/src/main/kotlin/com/passwordstore/android/ui/activity/SplashActivity.kt]: Will do once we implement DI, Dagger looks cool imo, we can try that.

File: app/src/main/kotlin/com/passwordstore/android/utils/Constants.kt (hunk 5 above)
Comment (Skrilltrax) [reply app/src/main/kotlin/com/passwordstore/android/utils/Constants.kt]: Sure

File: app/src/main/kotlin/com/passwordstore/android/ui/activity/SplashActivity.kt (hunk 6 above)
Comment (Skrilltrax) [reply app/src/main/kotlin/com/passwordstore/android/ui/activity/SplashActivity.kt]: That'll happen when we complete the first run flow. Changing here will complete the first run flow even before user chooses the repo location.

File: app/src/main/res/values/styles.xml (hunk 8 above)
Comment (Skrilltrax) [reply app/src/main/res/values/styles.xml]: We have seperate styles for v23, v27 and v29. This is present in all of those.

File: app/src/main/res/layout/activity_first_run.xml (hunk 7 above)
Comment (Skrilltrax) [reply app/src/main/res/layout/activity_first_run.xml]: Oh i messed up, will fix.

File: app/src/main/kotlin/com/passwordstore/android/ui/activity/FirstRunActivity.kt (hunk 3 above)
Comment (Skrilltrax) [reply app/src/main/kotlin/com/passwordstore/android/ui/activity/FirstRunActivity.kt]: :+1: Okay

File: app/src/main/res/values/styles.xml (hunk 8 above)
Comment (msfjarvis) [reply app/src/main/res/values/styles.xml]: :+1:

File: app/src/main/kotlin/com/passwordstore/android/ui/activity/SplashActivity.kt (hunk 1 above)
Comment (msfjarvis) [reply app/src/main/kotlin/com/passwordstore/android/ui/activity/SplashActivity.kt]: The empty `else` clause, unless something's supposed to go there that's gonna come later.

Comment (Skrilltrax) [reply app/src/main/kotlin/com/passwordstore/android/ui/activity/SplashActivity.kt]: We'll have code for syncing repos / other stuff we need to do to launch the MainActivity there.

File: app/src/main/kotlin/com/passwordstore/android/ui/activity/BaseActivity.kt (hunk 2 above)
Comment (Skrilltrax) [reply app/src/main/kotlin/com/passwordstore/android/ui/activity/BaseActivity.kt]: Fixing

**Comment by Skrilltrax** [comment]
Adding a few more things.

File: app/src/main/res/layout/activity_first_run.xml (hunk 7 above)
Comment (Skrilltrax) [reply app/src/main/res/layout/activity_first_run.xml]: Fixed

File: app/src/main/kotlin/com/passwordstore/android/utils/Extensions.kt [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+ */
+package com.passwordstore.android.utils
+
+import android.view.View
+import androidx.annotation.IdRes
+import androidx.fragment.app.Fragment
+import androidx.fragment.app.FragmentManager
+import com.passwordstore.android.R
+
+fun FragmentManager.performTransaction(destinationFragment: Fragment, @IdRes containerViewId: Int = android.R.id.content) {
```
Comment (msfjarvis) [inline app/src/main/kotlin/com/passwordstore/android/utils/Extensions.kt]: Add `fragment-ktx` dep and use the `commit` extensions for this file.

File: app/src/main/kotlin/com/passwordstore/android/utils/Extensions.kt [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+    this.beginTransaction()
+            .setCustomAnimations(R.animator.slide_in_left, R.animator.slide_out_left, R.animator.slide_in_right, R.animator.slide_out_right)
```
Comment (msfjarvis) [inline app/src/main/kotlin/com/passwordstore/android/utils/Extensions.kt]: Break this into multiple lines for readability

File: app/src/main/res/layout/activity_splash.xml [hunk 11]
```diff
@@ -5,7 +5,7 @@
     xmlns:app="http://schemas.android.com/apk/res-auto">
 
     <ImageView
```
Comment (msfjarvis) [inline app/src/main/res/layout/activity_splash.xml]: Make this `AppCompatImageView` for consistency?

File: app/src/main/res/layout/fragment_clone.xml [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+        android:id="@+id/iv_app_icon"
+        android:layout_width="64dp"
+        android:layout_height="64dp"
+        android:layout_marginStart="32dp"
+        android:layout_marginTop="100dp"
+        android:contentDescription="@string/app_icon_hint"
+        android:src="@mipmap/ic_launcher"
+        app:layout_constraintStart_toStartOf="parent"
+        app:layout_constraintTop_toTopOf="parent"
+        />
```
Comment (msfjarvis) [inline app/src/main/res/layout/fragment_clone.xml]: Move this to the previous line for consistency

File: app/src/main/res/layout/fragment_clone_bckup.xml [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+        android:layout_width="100dp"
+        android:layout_height="100dp"
+        android:src="@mipmap/ic_launcher"
+        android:contentDescription="@string/app_icon_hint"
+        android:transitionName="transition_first_app_icon"
+        app:layout_constraintTop_toTopOf="parent"
+        app:layout_constraintBottom_toBottomOf="parent"
+        app:layout_constraintStart_toStartOf="parent"
+        app:layout_constraintEnd_toEndOf="parent"
+        />
```
Comment (msfjarvis) [inline app/src/main/res/layout/fragment_clone_bckup.xml]: Move this to the previous line for consistency

File: app/src/main/res/layout/fragment_pgp_provider.xml [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+        android:id="@+id/iv_app_icon"
+        android:layout_width="64dp"
+        android:layout_height="64dp"
+        android:layout_marginStart="32dp"
+        android:layout_marginTop="100dp"
+        android:contentDescription="@string/app_icon_hint"
+        android:src="@mipmap/ic_launcher"
+        app:layout_constraintStart_toStartOf="parent"
+        app:layout_constraintTop_toTopOf="parent"
+        />
```
Comment (msfjarvis) [inline app/src/main/res/layout/fragment_pgp_provider.xml]: Move this to the previous line for consistency

File: app/src/main/res/layout/fragment_repo_location.xml [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+        android:id="@+id/iv_app_icon"
+        android:layout_width="64dp"
+        android:layout_height="64dp"
+        android:layout_marginStart="32dp"
+        android:layout_marginTop="100dp"
+        android:contentDescription="@string/app_icon_hint"
+        android:src="@mipmap/ic_launcher"
+        app:layout_constraintStart_toStartOf="parent"
+        app:layout_constraintTop_toTopOf="parent"
+        />
```
Comment (msfjarvis) [inline app/src/main/res/layout/fragment_repo_location.xml]: Move this to the previous line for consistency

File: app/src/main/res/layout/fragment_clone.xml (hunk 12 above)
Comment (Skrilltrax) [reply app/src/main/res/layout/fragment_clone.xml]: Done.

**Pull request merged**